### ISSUE TITLE
How to handle datetime arrays with timezone information?

### ISSUE THREAD

**Issue by weiji14** [GenericMappingTools/pygmt#3656]
_Originally posted by @weiji14 in https://github.com/GenericMappingTools/pygmt/pull/3621#discussion_r1861767477_

Opening a discussion on whether to:

1. Follow GMT (allow users to plot data at a non-UTC timezone, by ignoring the timezone offset)
2. Follow NumPy, whereby the data will from any non-UTC timezone will be converted to UTC always.

If going with 2, we should at least raise a warning if a non-UTC timezone is used, that a conversion is taking place. If going with 1, we would need to special-case datetime types, that might mean extra logic in the `_to_numpy` function, or having to keep `array_to_datetime` (so don't merge #3507 yet).

**Comment by seisman** [comment]
> datetime arrays (e.g. `pandas.Series` with `pandas.Timestamp`, `pyarrow.date32`, `pyarrow.date64`, etc) that have non-UTC timezones

`pyarrow.date32`/`pyarrow.date64` are timezone-unware. As far as I know, only `datetime.datetime`, `datetime.time`, `pandas.DatetimeTZDtype` and `pyarrow.timestamp` are timezone-aware.

---

I feel we should go with option 1, so the behavior is consistent with both GMT and the Python world (at least for matplotlib and pandas).

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

df = pd.DataFrame(
    {
        "datetime": pd.date_range(start="2024-01-01", periods=10, freq="H").tz_localize("Asia/Shanghai"), 
        "y": np.random.rand(10),
    }
) 
print(df["datetime"])

plt.figure(figsize=(10, 6))
plt.plot(df["datetime"], df["y"], marker="o", linestyle="-", color="blue", label="y values")
plt.title("Time Series Data (Hourly)")
plt.xlabel("Datetime (UTC+8)")
plt.ylabel("Y")
plt.show()

# Use pandas' plotting function
df.plot(
    x="datetime",
    y="y",
    kind="line",
    marker="o",
    title="Time Series Data (Hourly)",
    figsize=(10, 6),
    grid=True,
    xlabel="Datetime (UTC+8)",
    ylabel="Y Values",
    legend=True
)
plt.show()
```
![1](https://github.com/user-attachments/assets/2f3a6a70-d976-4e56-bfe7-5a520fefcbc3)
![2](https://github.com/user-attachments/assets/8fd1f849-d804-41bd-93b4-142e63c956e0)

**Comment by weiji14** [comment]
> > datetime arrays (e.g. `pandas.Series` with `pandas.Timestamp`, `pyarrow.date32`, `pyarrow.date64`, etc) that have non-UTC timezones
> 
> `pyarrow.date32`/`pyarrow.date64` are timezone-unware. As far as I know, only `datetime.datetime`, `datetime.time`, `pandas.DatetimeTZDtype` and `pyarrow.timestamp` are timezone-aware.

Yep, was typing this out too quickly. Have updated the top-post to clarify this.

---

> I feel we should go with option 1, so the behavior is consistent with both GMT and the Python world (at least for matplotlib and pandas).

I'm trying to get some historical context on how [matplotlib](https://github.com/search?q=repo%3Amatplotlib%2Fmatplotlib+plot+timezone&type=issues&s=&o=desc&p=3)/[pandas](https://github.com/search?q=repo%3Apandas-dev%2Fpandas+plot+timezone&type=code) has handled timezone plotting before. It seems like matplotlib used to have a [`plot_date()`](https://matplotlib.org/3.8.4/api/_as_gen/matplotlib.pyplot.plot_date.html) function which allows setting of a `tz` parameter, but this was deprecated in https://github.com/matplotlib/matplotlib/issues/18154 and it's unclear/confusing on what the regular matplotlib `.plot` function does now in terms of timezone handling.

There's also an interesting case on what to do when plotting two or more arrays with different timezones (e.g. UTC and Pacific time) at https://github.com/matplotlib/matplotlib/issues/8072#issuecomment-1919262892. With option 2, everything is converted to UTC so things will be plotted 'correctly'; but with option 1, ignoring the timezones means the data would be plotted in a non-deterministic manner depending on which array (UTC or Pacific) was passed in first.

**Comment by seisman** [comment]
> it's unclear/confusing on what the regular matplotlib `.plot` function does now in terms of timezone handling.

Isn't it clear that matplotlib's `plot` method is timezone-aware from my example above? The first point is plotted at 01-01 (Asia/Shanghai), not 12-31 (UTC).

> with option 1, ignoring the timezones means the data would be plotted in a non-deterministic manner depending on which array (UTC or Europe/Berlin) was passed in first.

This won't be a case in GMT or PyGMT, since arrays can be processed independently. What makes more troubles is the handling of datetime with different timezones in arguments like:
```
region = [
    datetime.datetime(2024, 1, 1, 10, 0, 0, tzinfo=zoneinfo.ZoneInfo('Asia/Shanghai')),
    datetime.datetime(2024, 1, 1, 10, 0, 0, tzinfo=zoneinfo.ZoneInfo('UTC')),
    0, 
    10
]
```
